### DISCUSSION TITLE
Docker swarm join

### DISCUSSION THREAD

**Discussion by Manu224781** [flatcar/Flatcar#1502] [Q&A] [answered]
Hey, everyone at Flatcar! Thanks for providing such an excellent operating system!

I'm currently in the process of deploying a substantial setup with Flatcar, and I need some best practice tips. What’s the best way to get the Docker process within Flatcar joined to a swarm?

My best guess is to create a systemd unit with an ExecStart to join the swarm. However, wouldn’t that run on every boot? What’s the best way to check if it is already joined?

If anyone is already running this setup, I’d be glad if you could share your experience.

Thanks!

**Comment by till** [comment, marked as answer]
So you either run something like Ansible outside the nodes or inspect the response from the join command? I don't know how granular it is since the node could be already joined, there could be transport errors and the token could be expired.

But maybe that gives you a hint as to if the node has joined.

I am still using Ansible for this part - basically do a docker node ls, parse and add/remove workers with a dynamic inventory.

I guess the systemd way would be to set a pre condition in the unit, e.g. a file shouldn't exist and then you execute the join command and create the file.

But I find that hacky and also leaves questions open for a DR (when managers break).

**Reply by till** [reply]
I have an ansible role that works with Flatcar and builds clusters. It's actually a fork of another but I yanked out Ubuntu/CentOS.

If this is of any interest, I can try to clean it up and publish it.

**Reply by Manu224781** [reply]
Hey @till that would be super helpful if you could do that. I really like the approach with ansible as well and will give that a go.

**Reply by till** [reply]
@Theni224 

I just tried to clean it up (so please ignore the git-history):
https://github.com/pngmbh/ansible-dockerswarm/pull/12

If you add something like this to your `requirements.yml` (for ansible-galaxy), it should work:

```
- name: atosatto.docker-swarm
  src: https://github.com/pngmbh/ansible-dockerswarm.git
  version: finish
```

... or download the branch and copy wherever you want to.

The Readme on the `finish` branch is updated and contains an example inventory. I'll see if I can merge/tag it this week or early next. We had been using an internal branch for the past 2 years or so 🙈 . 

Maybe as a caveat — we use a pretty ancient version of Ansible in-house (it's containerized to control Python and all the dependencies), mostly because we didn't want to deal with constant updates and BC breaks imposed by upstream. But this role is using very basic resources, so I hope it works with more recent versions too.

If anything you may have to update some resources, so this should get you started no matter what.

**Comment by tormath1** [comment]
Hi @Theni224 if you want to go ahead with a Systemd unit + ExecStart you can give a try to [`ConditionFirstBoot=true`](https://www.freedesktop.org/software/systemd/man/latest/systemd.unit.html#ConditionFirstBoot=) or as @till mentioned check that a file exists (or not).
